File: test/language/choice_types/cpp98/EnumParamChoiceTest.cpp
#include "gtest/gtest.h"

#include "choice_types/enum_param_choice/Selector.h"
#include "choice_types/enum_param_choice/EnumParamChoice.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace choice_types
{
namespace enum_param_choice
{

class EnumParamChoiceTest : public ::testing::Test
{
protected:
    void writeEnumParamChoiceToByteArray(zserio::BitStreamWriter& writer, Selector selector, int32_t value)
    {
        switch (selector.getValue())
        {
        case Selector::BLACK:
            writer.writeSignedBits(value, 8);
            break;

        case Selector::GREY:
            writer.writeSignedBits(value, 16);
            break;

        case Selector::WHITE:
            writer.writeSignedBits(value, 32);
            break;

        default:
            FAIL() << "Bad choice selector";
            break;
        }
    }
};

TEST_F(EnumParamChoiceTest, emptyConstructor)
{
    EnumParamChoice enumParamChoice;
    ASSERT_THROW(enumParamChoice.getSelector(), zserio::CppRuntimeException);
}

TEST_F(EnumParamChoiceTest, bitStreamReaderConstructor)
{
    const Selector selector = Selector::BLACK;
    const int8_t value = 99;
    zserio::BitStreamWriter writer;
    writeEnumParamChoiceToByteArray(writer, selector, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    EnumParamChoice enumParamChoice(reader, selector);
    ASSERT_EQ(selector, enumParamChoice.getSelector());
    ASSERT_EQ(value, enumParamChoice.getBlack());
}

TEST_F(EnumParamChoiceTest, copyConstructor)
{
    const Selector selector = Selector::BLACK;
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(selector);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);

    const EnumParamChoice enumParamChoiceCopy(enumParamChoice);
    ASSERT_EQ(selector, enumParamChoiceCopy.getSelector());
    ASSERT_EQ(value, enumParamChoiceCopy.getBlack());
}

TEST_F(EnumParamChoiceTest, operatorAssignment)
{
    const Selector selector = Selector::GREY;
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(selector);
    const int16_t value = 234;
    enumParamChoice.setGrey(value);

    const EnumParamChoice enumParamChoiceCopy = enumParamChoice;
    ASSERT_EQ(selector, enumParamChoiceCopy.getSelector());
    ASSERT_EQ(value, enumParamChoiceCopy.getGrey());
}

TEST_F(EnumParamChoiceTest, initialize)
{
    const Selector selector = Selector::GREY;
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(selector);
    ASSERT_EQ(selector, enumParamChoice.getSelector());
}

TEST_F(EnumParamChoiceTest, getSelector)
{
    const Selector selector = Selector::BLACK;
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(selector);
    ASSERT_EQ(selector, enumParamChoice.getSelector());
}

TEST_F(EnumParamChoiceTest, getSetBlack)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(Selector::BLACK);
    const int8_t value = 99;
    enumParamChoice.setBlack(value);
    ASSERT_EQ(value, enumParamChoice.getBlack());
}

TEST_F(EnumParamChoiceTest, getSetGrey)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(Selector::GREY);
    const int16_t value = 234;
    enumParamChoice.setGrey(value);
    ASSERT_EQ(value, enumParamChoice.getGrey());
}

TEST_F(EnumParamChoiceTest, getSetWhite)
{
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(Selector::WHITE);
    const int32_t value = 65535;
    enumParamChoice.setWhite(value);
    ASSERT_EQ(value, enumParamChoice.getWhite());
}

TEST_F(EnumParamChoiceTest, bitSizeOf)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(Selector::BLACK);
    ASSERT_EQ(8, enumParamChoiceB.bitSizeOf());

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(Selector::GREY);
    ASSERT_EQ(16, enumParamChoiceG.bitSizeOf());

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(Selector::WHITE);
    ASSERT_EQ(32, enumParamChoiceW.bitSizeOf());
}

TEST_F(EnumParamChoiceTest, initializeOffsets)
{
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(Selector::BLACK);
    const size_t bitPosition = 1;
    ASSERT_EQ(9, enumParamChoiceB.initializeOffsets(bitPosition));

    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(Selector::GREY);
    ASSERT_EQ(17, enumParamChoiceG.initializeOffsets(bitPosition));

    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(Selector::WHITE);
    ASSERT_EQ(33, enumParamChoiceW.initializeOffsets(bitPosition));
}

TEST_F(EnumParamChoiceTest, operatorEquality)
{
    EnumParamChoice enumParamChoice1;
    enumParamChoice1.initialize(Selector::BLACK);
    EnumParamChoice enumParamChoice2;
    enumParamChoice2.initialize(Selector::BLACK);
    ASSERT_TRUE(enumParamChoice1 == enumParamChoice2);

    const int8_t value = 99;
    enumParamChoice1.setBlack(value);
    ASSERT_FALSE(enumParamChoice1 == enumParamChoice2);

    enumParamChoice2.setBlack(value);
    ASSERT_TRUE(enumParamChoice1 == enumParamChoice2);

    const int8_t diffValue = value + 1;
    enumParamChoice2.setBlack(diffValue);
    ASSERT_FALSE(enumParamChoice1 == enumParamChoice2);
}

TEST_F(EnumParamChoiceTest, hashCode)
{
    EnumParamChoice enumParamChoice1;
    enumParamChoice1.initialize(Selector::BLACK);
    EnumParamChoice enumParamChoice2;
    enumParamChoice2.initialize(Selector::BLACK);
    ASSERT_EQ(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    const int8_t value = 99;
    enumParamChoice1.setBlack(value);
    ASSERT_NE(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    enumParamChoice2.setBlack(value);
    ASSERT_EQ(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

    const int8_t diffValue = value + 1;
    enumParamChoice2.setBlack(diffValue);
    ASSERT_NE(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());
}

TEST_F(EnumParamChoiceTest, read)
{
    const Selector selector = Selector::BLACK;
    EnumParamChoice enumParamChoice;
    enumParamChoice.initialize(selector);

    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeEnumParamChoiceToByteArray(writer, selector, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    enumParamChoice.read(reader);

    ASSERT_EQ(selector, enumParamChoice.getSelector());
    ASSERT_EQ(value, enumParamChoice.getBlack());
}

TEST_F(EnumParamChoiceTest, write)
{
    Selector selector = Selector::BLACK;
    EnumParamChoice enumParamChoiceB;
    enumParamChoiceB.initialize(selector);
    const int8_t valueB = 99;
    enumParamChoiceB.setBlack(valueB);
    zserio::BitStreamWriter writerB;
    enumParamChoiceB.write(writerB);
    size_t writeBufferByteSizeB;
    const uint8_t* writeBufferB = writerB.getWriteBuffer(writeBufferByteSizeB);
    zserio::BitStreamReader readerB(writeBufferB, writeBufferByteSizeB);
    EnumParamChoice readEnumParamChoiceB(readerB, selector);
    ASSERT_EQ(valueB, readEnumParamChoiceB.getBlack());

    selector = Selector::GREY;
    EnumParamChoice enumParamChoiceG;
    enumParamChoiceG.initialize(selector);
    const int16_t valueG = 234;
    enumParamChoiceG.setGrey(valueG);
    zserio::BitStreamWriter writerG;
    enumParamChoiceG.write(writerG);
    size_t writeBufferByteSizeG;
    const uint8_t* writeBufferG = writerG.getWriteBuffer(writeBufferByteSizeG);
    zserio::BitStreamReader readerG(writeBufferG, writeBufferByteSizeG);
    EnumParamChoice readEnumParamChoiceG(readerG, selector);
    ASSERT_EQ(valueG, readEnumParamChoiceG.getGrey());

    selector = Selector::WHITE;
    EnumParamChoice enumParamChoiceW;
    enumParamChoiceW.initialize(selector);
    const int32_t valueW = 65535;
    enumParamChoiceW.setWhite(valueW);
    zserio::BitStreamWriter writerW;
    enumParamChoiceW.write(writerW);
    size_t writeBufferByteSizeW;
    const uint8_t* writeBufferW = writerW.getWriteBuffer(writeBufferByteSizeW);
    zserio::BitStreamReader readerW(writeBufferW, writeBufferByteSizeW);
    EnumParamChoice readEnumParamChoiceW(readerW, selector);
    ASSERT_EQ(valueW, readEnumParamChoiceW.getWhite());
}

} // namespace enum_param_choice
} // namespace choice_types

File: test/language/choice_types/java/choice_types/EnumParamChoiceTest.java
package choice_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import choice_types.enum_param_choice.EnumParamChoice;
import choice_types.enum_param_choice.Selector;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class EnumParamChoiceTest
{
    @Test
    public void selectorConstructor()
    {
        final Selector selector = Selector.BLACK;
        final EnumParamChoice enumParamChoice = new EnumParamChoice(selector);
        assertEquals(selector, enumParamChoice.getSelector());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final Selector selector = Selector.GREY;
        final int value = 234;
        final BitBuffer buffer = writeEnumParamChoiceToBitBuffer(selector, value);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(buffer.getBuffer(),
                buffer.getBitSize());
        final EnumParamChoice enumParamChoice = new EnumParamChoice(reader, selector);
        assertEquals(selector, enumParamChoice.getSelector());
        assertEquals((short)value, enumParamChoice.getGrey());
    }

    @Test
    public void choiceTag()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.BLACK);
        assertEquals(EnumParamChoice.CHOICE_black, enumParamChoice.choiceTag());

        enumParamChoice = new EnumParamChoice(Selector.GREY);
        assertEquals(EnumParamChoice.CHOICE_grey, enumParamChoice.choiceTag());

        enumParamChoice = new EnumParamChoice(Selector.WHITE);
        assertEquals(EnumParamChoice.CHOICE_white, enumParamChoice.choiceTag());
    }

    @Test
    public void bitSizeOf()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.BLACK);
        assertEquals(8, enumParamChoice.bitSizeOf());

        enumParamChoice = new EnumParamChoice(Selector.GREY);
        assertEquals(16, enumParamChoice.bitSizeOf());

        enumParamChoice = new EnumParamChoice(Selector.WHITE);
        assertEquals(32, enumParamChoice.bitSizeOf());
    }

    @Test
    public void getSelector()
    {
        final Selector selector = Selector.BLACK;
        final EnumParamChoice enumParamChoice = new EnumParamChoice(selector);
        assertEquals(selector, enumParamChoice.getSelector());
    }

    @Test
    public void getSetBlack()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.BLACK);
        final byte value = 99;
        enumParamChoice.setBlack(value);
        assertEquals(value, enumParamChoice.getBlack());
    }

    @Test
    public void getSetGrey()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.GREY);
        final short value = 234;
        enumParamChoice.setGrey(value);
        assertEquals(value, enumParamChoice.getGrey());
    }

    @Test
    public void getSetWhite()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.WHITE);
        final int value = 65535;
        enumParamChoice.setWhite(value);
        assertEquals(value, enumParamChoice.getWhite());
    }

    @Test
    public void equals()
    {
        EnumParamChoice enumParamChoice1 = new EnumParamChoice(Selector.BLACK);
        EnumParamChoice enumParamChoice2 = new EnumParamChoice(Selector.BLACK);
        assertTrue(enumParamChoice1.equals(enumParamChoice2));

        final byte value = 99;
        enumParamChoice1.setBlack(value);
        assertFalse(enumParamChoice1.equals(enumParamChoice2));

        enumParamChoice2.setBlack(value);
        assertTrue(enumParamChoice1.equals(enumParamChoice2));

        final byte diffValue = value + 1;
        enumParamChoice2.setBlack(diffValue);
        assertFalse(enumParamChoice1.equals(enumParamChoice2));
    }

    @Test
    public void hashCodeMethod()
    {
        EnumParamChoice enumParamChoice1 = new EnumParamChoice(Selector.BLACK);
        EnumParamChoice enumParamChoice2 = new EnumParamChoice(Selector.BLACK);
        assertEquals(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

        final byte value = 99;
        enumParamChoice1.setBlack(value);
        assertTrue(enumParamChoice1.hashCode() != enumParamChoice2.hashCode());

        enumParamChoice2.setBlack(value);
        assertEquals(enumParamChoice1.hashCode(), enumParamChoice2.hashCode());

        final byte diffValue = value + 1;
        enumParamChoice2.setBlack(diffValue);
        assertTrue(enumParamChoice1.hashCode() != enumParamChoice2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(63073, enumParamChoice1.hashCode());
        assertEquals(63074, enumParamChoice2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        EnumParamChoice enumParamChoice = new EnumParamChoice(Selector.BLACK);
        final int bitPosition = 1;
        assertEquals(9, enumParamChoice.initializeOffsets(bitPosition));

        enumParamChoice = new EnumParamChoice(Selector.GREY);
        assertEquals(17, enumParamChoice.initializeOffsets(bitPosition));

        enumParamChoice = new EnumParamChoice(Selector.WHITE);
        assertEquals(33, enumParamChoice.initializeOffsets(bitPosition));
    }

    @Test
    public void bitStreamWriterWrite() throws IOException, ZserioError
    {
        Selector selector = Selector.BLACK;
        EnumParamChoice enumParamChoice = new EnumParamChoice(selector);
        final byte byteValue = 99;
        enumParamChoice.setBlack(byteValue);
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        enumParamChoice.write(writer);
        ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray(),
                writer.getBitPosition());
        EnumParamChoice readEnumParamChoice = new EnumParamChoice(reader, selector);
        assertEquals(byteValue, readEnumParamChoice.getBlack());

        selector = Selector.GREY;
        enumParamChoice = new EnumParamChoice(selector);
        final short shortValue = 234;
        enumParamChoice.setGrey(shortValue);
        writer = new ByteArrayBitStreamWriter();
        enumParamChoice.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        readEnumParamChoice = new EnumParamChoice(reader, selector);
        assertEquals(shortValue, readEnumParamChoice.getGrey());

        selector = Selector.WHITE;
        enumParamChoice = new EnumParamChoice(selector);
        final int intValue = 65535;
        enumParamChoice.setWhite(intValue);
        writer = new ByteArrayBitStreamWriter();
        enumParamChoice.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        readEnumParamChoice = new EnumParamChoice(reader, selector);
        assertEquals(intValue, readEnumParamChoice.getWhite());
    }

    private BitBuffer writeEnumParamChoiceToBitBuffer(Selector selector, int value) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            if (selector == Selector.BLACK)
                writer.writeByte((byte)value);
            else if (selector == Selector.GREY)
                writer.writeShort((short)value);
            else if (selector == Selector.WHITE)
                writer.writeInt(value);
            else
                fail("Invalid selector: " + selector);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }
}
